File: tests/metrics/tracking_error_test.py
from __future__ import annotations

import numpy as np
import pytest

from taflow.metrics.tracking_error import TrackingError


def expected_tracking_error(
    returns: np.ndarray,
    benchmark_returns: np.ndarray,
    periods_per_year: float,
    annualized: bool,
) -> float:
    """Independent NumPy oracle with explicit pairwise missing omission."""
    usable = ~(np.isnan(returns) | np.isnan(benchmark_returns))
    scale = np.sqrt(periods_per_year) if annualized else 1.0
    return float(np.std(returns[usable] - benchmark_returns[usable], ddof=1) * scale)


@pytest.mark.parametrize("annualized", [False, True])
@pytest.mark.parametrize("periods_per_year", [1.0, 12.0, 252.0, 8760.0])
def test_tracking_error_matches_numpy(
    annualized: bool, periods_per_year: float
) -> None:
    returns = np.array([0.01, np.nan, -0.03, 0.02, 0.005, -0.01])
    benchmark = np.array([0.005, 0.01, -0.01, np.nan, 0.002, -0.02])
    actual = TrackingError.from_returns(
        returns,
        benchmark,
        periods_per_year=periods_per_year,
        annualized=annualized,
    ).compute()
    expected = expected_tracking_error(
        returns, benchmark, periods_per_year, annualized
    )
    assert actual == pytest.approx(expected, rel=1e-13, abs=1e-15)


def test_tracking_error_random_constant_and_minimum_cases() -> None:
    random = np.random.default_rng(20260811)
    benchmark = random.normal(0.0003, 0.01, 513)
    returns = benchmark + random.normal(0.0001, 0.004, 513)
    expected = expected_tracking_error(returns, benchmark, 252.0, True)
    assert TrackingError.from_returns(returns, benchmark).compute() == pytest.approx(
        expected, rel=1e-13, abs=1e-15
    )
    assert TrackingError.from_returns([0.01], [0.00]).compute() is None
    assert TrackingError.from_returns([0.02, 0.02], [0.01, 0.01]).compute() == 0.0


def test_tracking_error_factories_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.10, -0.20, 0.05])
    benchmark_returns = np.array([0.02, -0.10, 0.01])
    expected = TrackingError.from_returns(
        returns, benchmark_returns, periods_per_year=12.0
    ).compute()
    equity = np.array([100.0, 110.0, 88.0, 92.4])
    benchmark_equity = np.array([200.0, 204.0, 183.6, 185.436])

    assert TrackingError.from_equity(
        equity, benchmark_equity, periods_per_year=12.0
    ).compute() == pytest.approx(expected)
    assert TrackingError.from_pnl(
        np.array([10.0, -22.0, 4.4]),
        np.array([4.0, -20.4, 1.836]),
        initial_equity=100.0,
        benchmark_initial_equity=200.0,
        periods_per_year=12.0,
    ).compute() == pytest.approx(expected)
    assert TrackingError.from_log_returns(
        np.log1p(returns),
        np.log1p(benchmark_returns),
        periods_per_year=12.0,
    ).compute() == pytest.approx(expected)

    state = TrackingError.from_returns([], [], periods_per_year=12.0)
    assert state.value is None
    assert state.append(returns[0], benchmark_returns[0]) is state
    assert state.value is None
    assert state.extend(returns[1:], benchmark_returns[1:]) is state
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(returns)
    assert state.reset() is state
    assert len(state) == 0
    assert state.extend(returns, benchmark_returns).compute() == pytest.approx(expected)


def test_tracking_error_rejects_misalignment_without_mutation() -> None:
    state = TrackingError.from_returns([0.01, 0.02], [0.00, 0.01])
    original_value = state.value
    original_length = len(state)
    with pytest.raises(ValueError, match="equal length"):
        state.extend([0.03, 0.04], [0.02])
    assert len(state) == original_length
    assert state.value == original_value


def test_tracking_error_missing_and_invalid_values() -> None:
    state = TrackingError.from_returns(
        [np.nan, 0.01, 0.02], [0.00, 0.00, np.nan]
    )
    assert len(state) == 1
    assert state.compute() is None
    with pytest.raises(ValueError):
        TrackingError.from_returns([np.nan], [0.0], nan_policy="raise")
    with pytest.raises(ValueError):
        TrackingError.from_returns([np.inf], [0.0])
    with pytest.raises(ValueError):
        TrackingError.from_returns([0.01, 0.02], [0.01])
    with pytest.raises(ValueError):
        TrackingError.from_returns([0.01, 0.02], [0.00, 0.01], periods_per_year=0.0)


def test_tracking_error_requires_semantic_factory() -> None:
    with pytest.raises(TypeError):
        TrackingError()

File: crates/taflow-metrics/src/metrics/tracking_error_test.rs
use super::tracking_error::TrackingError;
use crate::{MetricInputKind, NanPolicy};

fn assert_close(actual: f64, expected: f64) {
    assert!((actual - expected).abs() <= 1e-14, "{actual} != {expected}");
}

#[test]
fn computes_annualized_sample_active_return_deviation() {
    let primary = [0.03, -0.01, 0.02, 0.04];
    let benchmark = [0.01, -0.02, 0.025, 0.01];
    let active = [0.02, 0.01, -0.005, 0.03];
    let mean = active.iter().sum::<f64>() / active.len() as f64;
    let sample_variance = active
        .iter()
        .map(|value| (value - mean).powi(2))
        .sum::<f64>()
        / (active.len() - 1) as f64;
    let expected = sample_variance.sqrt() * 252.0_f64.sqrt();
    let mut state = TrackingError::new(252.0, true, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[], &[])?;
            Ok(state)
        })
        .unwrap();

    assert_eq!(state.value(), None);
    assert_eq!(state.append(primary[0], benchmark[0]).unwrap(), None);
    state.extend(&primary[1..3], &benchmark[1..3]).unwrap();
    assert_close(
        state.append(primary[3], benchmark[3]).unwrap().unwrap(),
        expected,
    );
    assert_close(state.compute().unwrap(), expected);
    assert_eq!(state.len(), primary.len());

    state.reset();
    assert!(state.is_empty());
    assert_eq!(state.value(), None);
    assert_close(
        state.extend(&primary, &benchmark).unwrap().unwrap(),
        expected,
    );
}

#[test]
fn supports_unannualized_output_and_pairwise_missing_omission() {
    let mut state = TrackingError::new(12.0, false, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[], &[])?;
            Ok(state)
        })
        .unwrap();
    state
        .extend(
            &[0.01, f64::NAN, 0.05, -0.02],
            &[0.00, 0.02, f64::NAN, -0.01],
        )
        .unwrap();

    let expected = 0.0002_f64.sqrt();
    assert_close(state.value().unwrap(), expected);
    assert_eq!(state.len(), 2);
}

#[test]
fn input_modes_produce_equivalent_tracking_error() {
    let primary_returns = [0.10, -0.20, 0.05];
    let benchmark_returns = [0.02, -0.10, 0.01];
    let expected = {
        let mut state = TrackingError::new(12.0, true, NanPolicy::Omit)
            .and_then(|mut state| {
                state.from_returns(&[], &[])?;
                Ok(state)
            })
            .unwrap();
        state
            .extend(&primary_returns, &benchmark_returns)
            .unwrap()
            .unwrap()
    };

    let mut equity = TrackingError::new(12.0, true, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_equity(&[], &[])?;
            Ok(state)
        })
        .unwrap();
    assert_close(
        equity
            .extend(&[100.0, 110.0, 88.0, 92.4], &[200.0, 204.0, 183.6, 185.436])
            .unwrap()
            .unwrap(),
        expected,
    );

    let mut pnl = TrackingError::new(12.0, true, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_pnl(&[], &[], 100.0, 200.0)?;
            Ok(state)
        })
        .unwrap();
    assert_close(
        pnl.extend(&[10.0, -22.0, 4.4], &[4.0, -20.4, 1.836])
            .unwrap()
            .unwrap(),
        expected,
    );

    let mut logarithmic = TrackingError::new(12.0, true, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_log_returns(&[], &[])?;
            Ok(state)
        })
        .unwrap();
    let primary_log_returns = primary_returns.map(f64::ln_1p);
    let benchmark_log_returns = benchmark_returns.map(f64::ln_1p);
    assert_close(
        logarithmic
            .extend(&primary_log_returns, &benchmark_log_returns)
            .unwrap()
            .unwrap(),
        expected,
    );
}

#[test]
fn rejects_misalignment_invalid_configuration_and_domains() {
    let mut state = TrackingError::new(252.0, true, NanPolicy::Omit)
        .and_then(|mut state| {
            state.from_returns(&[], &[])?;
            Ok(state)
        })
        .unwrap();
    assert!(state.extend(&[0.01, 0.02], &[0.01]).is_err());
    assert_eq!(state.len(), 0);

    for periods_per_year in [0.0, -1.0, f64::NAN, f64::INFINITY] {
        assert!(TrackingError::new(periods_per_year, true, NanPolicy::Omit)
            .and_then(|mut state| {
                state.from_returns(&[], &[])?;
                Ok(state)
            })
            .is_err());
    }
    assert!(TrackingError::new(252.0, true, NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0, 0.0)?;
            Ok(state)
        })
        .is_err());
    assert!(TrackingError::new(252.0, true, NanPolicy::Omit)
        .and_then(|mut state| {
            state.append(0.0, 0.0)?;
            Ok(state)
        })
        .is_err());
}
